### DISCUSSION TITLE
Using call/2 to send messages

### DISCUSSION THREAD

**Discussion by hnbeck** [LogtalkDotOrg/logtalk3#116] [Q&A]
Hi, 

what is not clear to me: I can do things in Prolog like
`A = myPredicate,
B = myArgument,
call(A, B).
`
I can do in Logtalk
`A = myObject
B=myMessage,
A::B.`

if myMessage/0.  But what is the way to do it if myMessage requires Arguments (myMessage/N, N>0)?

Cheers

Hans

**Comment by pmoura** [comment]
Isn't doing the following practical in your case:

```logtalk
..., Message =.. [Name| Arguments], Object::Message, ...
```
In alternative, **if** the goal is called from inside an object (or category), you can simply do:

```logtalk
..., call(Object::Name, Arg1, ...), ...
```

In Logtalk, `call/N` are interpreted as **private** meta-predicates and thus cannot be used as messages. The main reason for the private scope is that meta-arguments are always called in the context of the *sender*. Thus, a `Object::call(Name, Arg1, ...)` goal would imply calling `Name(Arg1, ...)` in *sender* and not being interpreted as a message to the object.

**Reply by hnbeck** [reply]
Hi Paulo, 

argh, yes of course :). That's perfect, I forgot this possibility. Sorry for the trivial question :)

Cheers

Hans

**Reply by pmoura** [reply]
Not really trivial as far as `call/N` handling of closures is concerned. Outside an object (or category), `call/N` is Prolog `call/N` and thus doesn't handle properly closures that use Logtalk control constructs such as, for example, `::/1`, `::/2`, or `^^/1`. But inside an object (or category), `call/N` is handled by the Logtalk compiler/runtime and is thus aware of all Logtalk control constructs. Then there's also the difference between Prolog and Logtalk meta-predicate semantics. While in Logtalk meta-arguments are always called in the calling context, in Prolog meta-arguments are called in either the calling context or in the meta-predicate definition context depending if the call is, respectively, implicitly or explicitly module-qualified.
